datalad_registry/tasks/__init__.py

```diff
@@ -0,0 +1,556 @@
+
+    dataset_url.tags = json.dumps(ds.repo.get_tags())
+
+    dataset_url.git_objects_kb = (
+        ds.repo.count_objects["size"] + ds.repo.count_objects["size-pack"]
+    )
+
+    dataset_url.last_update_dt = datetime.now(timezone.utc)
+
+
+# Map of extractors to their respective required files
+#     The required files are specified relative to the root of the dataset
+_EXTRACTOR_REQUIRED_FILES = {
+    "metalad_studyminimeta": [".studyminimeta.yaml"],
+    "datacite_gin": ["datacite.yml"],
+    "bids_dataset": ["dataset_description.json"],
+}
+
+
+@shared_task
+def log_error(request, exc, traceback) -> None:
+    """
+    An error handler for logging errors in tasks
+    :param request: The request in fulfilling which the error/exception has occurred
+    :param exc: The exception that has occurred
+    :param traceback: The traceback of the exception
+    """
+    lgr.error("%s\n%r\n%s\n", request, exc, traceback)
+
+
+# `acks_late` is set. Make sure this task is always idempotent
+@shared_task(acks_late=True)
+@validate_arguments
+def extract_ds_meta(ds_url_id: StrictInt, extractor: StrictStr) -> ExtractMetaStatus:
+    """
+    Extract dataset level metadata from a dataset
+
+    :param ds_url_id: The ID (primary key) of the RepoUrl of the dataset in the database
+    :param extractor: The name of the extractor to use
+    :return: `ExtractMetaStatus.SUCCEEDED` if the extraction has produced
+                 valid metadata. In this case, the metadata has been recorded to
+                 the database upon return.
+             `ExtractMetaStatus.ABORTED` if the extraction has been aborted due to some
+                 required files not being present in the dataset. For example,
+                 `.studyminimeta.yaml` is not present in the dataset for running
+                 the studyminimeta extractor.
+             `ExtractMetaStatus.SKIPPED` if the extraction has been skipped because the
+                 metadata to be extracted is already present in the database,
+                 as identified by the extractor name, RepoUrl, and dataset version.
+             `ExtractMetaStatus.NO_RECORD` if there is no RepoUrl in the database with
+                the specified ID. (This task can be initiated with the argument of a
+                supposed ID of a RepoUrl that doesn't identify any RepoUrl in
+                the database at the time of the execution of this task because
+                the RepoUrl has been deleted from the database.)
+    :raise: ValueError if the RepoUrl of the specified ID does not exist or has not
+                been processed yet.
+            RuntimeError if the extraction has produced no valid metadata.
+
+    """
+
+    # Get the RepoUrl from the database by ID with a read/share lock
+    url = (
+        db.session.execute(
+            select(RepoUrl).filter_by(id=ds_url_id).with_for_update(read=True)
+        )
+        .scalars()
+        .one_or_none()
+    )
```

Comment:
smells like having some helper which absorbs this pattern might be worthwhile (not in this PR, generally)

```python
url = get_url(ds_url_id, read=True, none_ok=True)
```